cogeo_mosaic/backend/base.py

```diff
@@ -0,0 +1,70 @@
+"""cogeo_mosaic.backend.base: base Backend class."""
+
+import abc
+from contextlib import AbstractContextManager
+from typing import Dict, Optional
+
+from cogeo_mosaic import version as mosaic_version
+from cogeo_mosaic.backend.utils import get_hash
+from cogeo_mosaic.utils import create_mosaic
+
+
+class BaseBackend(AbstractContextManager):
+    mosaic_def: Optional[Dict]
+
+    @abc.abstractmethod
+    def __init__(self, *args, **kwargs):
+        """Connect to backend"""
```

Comment:
Connecting to any resources required to access the backend should happen within `__enter__` and cleaned up within `__exit__`

Reply:
I guess I just thought it was simpler to do it inside `__init__` ... Is there ever a time when `__init__` is called but not `__enter__`?

Reply:
Yeah `__init__` is called whenever an object is created.  `__enter__` is only called when the object is entered using a `with` statement.

Reply:
I guess I wasn't sure that I wanted to force always using a context manager. Like when you put necessary logic in `__init__`, it means that 

1. Someone can choose not to use a context manager if they don't want to
2. There's never an occasion when a class variable is expected to be defined, but is only defined in `__enter__` instead of `__init__` so code unexpectedly breaks

Reply:
I don't think it really matters if we use the backends as context managers or not, as long as we are consistent (so the call pattern across backends is the same) and the usage is clear from the base class.  Inheriting from `AbstractContextManager` makes me think that backends are supposed to be used with context.

Reply:
Ok, I guess I don't have super strong feelings, but let me know if you think I should change something